File: 2025/146. LRU Cache/Solution.java
class Node {
    int key;
    int value;
    Node prev;
    Node next;
    public Node(int key, int value, Node prev, Node next) {
        this.key = key;
        this.value = value;
        this.prev = prev;
        this.next = next;
    }
}

class LRUCache {

    Map<Integer, Node> map;
    Node head, tail;
    int len, capacity;

    public LRUCache(int capacity) {
        this.map = new HashMap<>();
        this.head = new Node(-1, -1, null, null);
        this.tail = head;
        this.capacity = capacity;
    }
    
    public int get(int key) {
        int res = -1;
        if (map.containsKey(key)) {
            Node n = map.get(key);
            moveNode(n);
            return n.value;
        }
        return res;
    }
    
    public void put(int key, int value) {
        if (map.containsKey(key)) {
            Node n = map.get(key);
            n.value = value;
            moveNode(n);
        } else {
            Node n = new Node(key, value, tail, null);
            tail.next = n;
            tail = tail.next;
            tail.next = null;
            map.put(key, n);
            evictNode();
        }
    }

    private void moveNode(Node n) {
        if (tail != n) {
            n.next.prev = n.prev; 
            n.prev.next = n.next; 
            tail.next = n;
            n.prev = tail;
            tail = tail.next;
            tail.next = null;   
        }
    }

    private void evictNode() {
        if (map.size() > capacity) {
            map.remove(head.next.key);
            head.next.next.prev = head;
            head.next = head.next.next;
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 2025/146. LRU Cache/solution.py
class LRUNode:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.count = 0
        self.capacity = capacity
        self.key_map = {}
        self.head, self.tail = LRUNode(-1, -1), LRUNode(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key not in self.key_map:
            return -1
        node = self.key_map[key]
        self._move_node(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        if key not in self.key_map:
            node = LRUNode(key, value)
            self.key_map[key] = node
            self.count += 1
            self._add_node(node)
            if self.count > self.capacity:
                self.evict()
        else:
            node = self.key_map[key]
            node.value = value    
            self._move_node(node)

    def evict(self) -> None:
        remove_node = self.tail.prev
        remove_node.prev.next, self.tail.prev = self.tail, remove_node.prev
        del self.key_map[remove_node.key]
        del remove_node

    def _add_node(self, node) -> None:
        node.next, self.head.next.prev = self.head.next, node
        self.head.next, node.prev = node, self.head

    def _move_node(self, node) -> None:
        node.prev.next, node.next.prev = node.next, node.prev
        self._add_node(node)
        
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)